dali/operators/segmentation/select_masks.cc

```diff
@@ -0,0 +1,199 @@
+               make_string("``mask_coords`` input is expected to be 2D. Got ",
+                           in_mask_coords_shape.sample_dim(), "D"));
+
+  int nsamples = in_mask_meta.size();
+  DALI_ENFORCE(nsamples == in_mask_meta_shape.size() && nsamples == in_mask_coords_shape.size(),
+               make_string("All the inputs should have the same number of samples. Got: ", nsamples,
+                           ", ", in_mask_meta_shape.size(), ", ", in_mask_coords_shape.size()));
+  assert(nsamples > 0);
+
+  for (int i = 0; i < nsamples; i++) {
+    DALI_ENFORCE(3 == in_mask_meta_shape[i][1],
+                 make_string("``mask_meta`` is expected to contain 3 element rows, containing "
+                             "``mask_id, start_idx, end_idx``. Got ",
+                             in_mask_meta_shape[i][1], " elements"));
+  }
+
+  int coord_ndim = in_mask_coords_shape[0][1];
+  for (int i = 1; i < nsamples; i++) {
+    DALI_ENFORCE(
+        coord_ndim == in_mask_coords_shape[i][1],
+        make_string("All coordinates are expected to have the same dimensionality. Got ",
+                    coord_ndim, "D and ", in_mask_coords_shape[i][1], "D in the same batch"));
+  }
+
+  const auto &in_mask_ids_view = view<const int32_t, 1>(in_mask_ids);
+  const auto &in_mask_meta_view = view<const int32_t, 2>(in_mask_meta);
+  const auto &in_mask_coords_view = view<const float, 2>(in_mask_coords);
+
+  auto out_mask_meta_shape = in_mask_meta_shape;
+  auto out_mask_coords_shape = in_mask_coords_shape;
+
+  samples_meta_.resize(nsamples);
+  for (int i = 0; i < nsamples; i++) {
+    auto &meta = samples_meta_[i];
+    meta.clear();
+    int64_t nselected = in_mask_ids_view.tensor_shape_span(i)[0];
+    meta.selected_masks = make_cspan(in_mask_ids_view.tensor_data(i), nselected);
+    out_mask_meta_shape.tensor_shape_span(i)[0] = meta.selected_masks.size();
+    int idx = 0;
+    for (auto mask_id : meta.selected_masks) {
+      if (meta.masks_meta.find(mask_id) != meta.masks_meta.end()) {
+        DALI_WARN(make_string("mask_id ", mask_id, " is duplicated. Ignoring..."));
+        continue;
+      }
+      meta.masks_meta[mask_id].mask_id = mask_id;
+      meta.masks_meta[mask_id].new_mask_id = reindex_masks_ ? idx++ : mask_id;
+    }
+
+    int64_t nmasks = in_mask_meta_shape.tensor_shape_span(i)[0];
+    for (int64_t k = 0; k < nmasks; k++) {
+      const auto *mask_data = in_mask_meta_view.tensor_data(i) + k * 3;
+      int mask_id = mask_data[0];
+      auto it = meta.masks_meta.find(mask_id);
+      if (it == meta.masks_meta.end())
+        continue;
+      it->second.start_coord = mask_data[1];
+      it->second.end_coord = mask_data[2];
+    }
+
+    int64_t ncoords = 0;
+    for (int k = 0; k < nselected; k++) {
+      int mask_id = meta.selected_masks[k];
+      auto it = meta.masks_meta.find(mask_id);
+      if (it == meta.masks_meta.end()) {
+        DALI_FAIL(make_string("Selected mask_id ", mask_id,
```

Comment:
Shouldn't that FAIL be moved to L136?
I don't think that is possible here as:
```
 for (auto mask_id : meta.selected_masks) {
      if (meta.masks_meta.find(mask_id) != meta.masks_meta.end()) {
        DALI_WARN(make_string("mask_id ", mask_id, " is duplicated. Ignoring..."));
        continue;
      }
      meta.masks_meta[mask_id].mask_id = mask_id;
      meta.masks_meta[mask_id].new_mask_id = reindex_masks_ ? idx++ : mask_id;
    }
```
So instead of range for, you have 0 to nselected for, but sill mask_ids are the same.